getTracer() fails when the UDP socket hostname isn't resolvable

We had this backtrace recently when a developer had a bad minikube configuration.
```Exception in thread "main" java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
    at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
    at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
    at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.lang.RuntimeException: TUDPTransport cannot connect:
    at com.uber.jaeger.reporters.protocols.ThriftUdpTransport.newThriftUdpClient(ThriftUdpTransport.java:57)
    at com.uber.jaeger.senders.UdpSender.<init>(UdpSender.java:54)
    at com.uber.jaeger.Configuration$ReporterConfiguration.getSender(Configuration.java:328)
    at com.uber.jaeger.Configuration$ReporterConfiguration.getReporter(Configuration.java:335)
    at com.uber.jaeger.Configuration$ReporterConfiguration.access$000(Configuration.java:282)
    at com.uber.jaeger.Configuration.getTracerBuilder(Configuration.java:162)
    at com.uber.jaeger.Configuration.getTracer(Configuration.java:172)
...
Caused by: java.net.SocketException: Unresolved address
    at java.net.DatagramSocket.connect(DatagramSocket.java:493)
    at com.uber.jaeger.reporters.protocols.ThriftUdpTransport.newThriftUdpClient(ThriftUdpTransport.java:55)
    ... 19 more
```

Now, for production contexts I can get behind 'broken configs should stop the program', but for developer contexts logging a message would be quite ok. A further aspect is whether the brokenness is actually broken, or transient.

For UDP protocols we cannot tell if the other end is running or not, so we don't know if the endpoint address and port are correct or not in general. Stopping therefore when they are broken in a specific way doesn't seem useful to me: we have to depend on other mechanisms to determine correctness in prod *no matter what*. That makes this qualitatively different to e.g. bad sampler configurations.

e.g. It might be nice to not throw RuntimeError for this at all, and just log it.

--- a/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/UdpSender.java
+++ b/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/UdpSender.java
@@ -27,8 +27,11 @@ public class UdpSender extends ThriftSender {
   public static final String DEFAULT_AGENT_UDP_HOST = "localhost";
   public static final int DEFAULT_AGENT_UDP_COMPACT_PORT = 6831;
 
-  @ToString.Exclude private Agent.Client agentClient;
-  @ToString.Exclude private ThriftUdpTransport udpTransport;
+  private final String host;
+  private final int port;
+
+  @ToString.Exclude private volatile Agent.Client agentClient;
+  @ToString.Exclude private volatile ThriftUdpTransport udpTransport;
 
   /**
    * This constructor expects Jaeger running running on {@value #DEFAULT_AGENT_UDP_HOST}
@@ -54,14 +57,30 @@ public class UdpSender extends ThriftSender {
       port = DEFAULT_AGENT_UDP_COMPACT_PORT;
     }
 
-    udpTransport = ThriftUdpTransport.newThriftUdpClient(host, port);
-    agentClient = new Agent.Client(protocolFactory.getProtocol(udpTransport));
+    this.host = host;
+    this.port = port;
+  }
+
+  private Agent.Client getAgentClient() {
+    Agent.Client localRef = this.agentClient;
+    if (localRef == null) {
+      synchronized (this) {
+        localRef = this.agentClient;
+        if (localRef == null) {
+          udpTransport = ThriftUdpTransport.newThriftUdpClient(host, port);
+          localRef = new Agent.Client(protocolFactory.getProtocol(udpTransport));
+          this.agentClient = localRef;
+        }
+      }
+    }
+
+    return localRef;
   }
 
   @Override
   public void send(Process process, List<io.jaegertracing.thriftjava.Span> spans) throws SenderException {
     try {
-      agentClient.emitBatch(new Batch(process, spans));
+      getAgentClient().emitBatch(new Batch(process, spans));
     } catch (Exception e) {
       throw new SenderException(String.format("Could not send %d spans", spans.size()), e, spans.size());
     }
@@ -72,7 +91,11 @@ public class UdpSender extends ThriftSender {
     try {
       return super.close();
     } finally {
-      udpTransport.close();
+      synchronized (this) {
+        if (udpTransport != null) {
+          udpTransport.close();
+        }
+      }
     }
   }
 }

--- a/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/ThriftSenderFactoryTest.java
+++ b/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/ThriftSenderFactoryTest.java
@@ -17,7 +17,6 @@ package io.jaegertracing.thrift.internal.senders;
 import static org.junit.Assert.assertTrue;
 
 import io.jaegertracing.Configuration;
-import io.jaegertracing.internal.senders.NoopSender;
 import io.jaegertracing.spi.Sender;
 import org.junit.Before;
 import org.junit.Test;
@@ -47,7 +46,7 @@ public class ThriftSenderFactoryTest {
     System.setProperty(Configuration.JAEGER_AGENT_HOST, "jaeger-agent");
     System.setProperty(Configuration.JAEGER_AGENT_PORT, "6832");
     Sender sender = Configuration.SenderConfiguration.fromEnv().getSender();
-    assertTrue(sender instanceof NoopSender);
+    assertTrue(sender instanceof UdpSender);
   }
 
   @Test

--- a/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpLazinessTest.java
+++ b/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpLazinessTest.java
@@ -1,0 +1,37 @@
+/*
+ * Copyright (c) 2020, The Jaeger Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.jaegertracing.thrift.internal.senders;
+
+import io.jaegertracing.internal.exceptions.SenderException;
+import org.junit.Test;
+
+import java.net.SocketException;
+import java.util.Collections;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class UdpLazinessTest {
+    @Test
+    public void testLazyInitializationOfAgent() {
+        UdpSender udpSender = new UdpSender("agent.acme.test", 55555, 0);
+
+        try {
+            udpSender.send(null, Collections.emptyList());
+            fail("Send should fail!");
+        } catch (SenderException e) {
+            assertTrue("send should throw a socket exception", e.getCause().getCause() instanceof SocketException);
+        }
+    }
+}

